lib/web/join_tokens.go

```diff
@@ -240,20 +241,35 @@ func createJoinToken(ctx context.Context, m nodeAPIGetter, roles types.SystemRol
 }
 
 func getJoinScript(settings scriptSettings, m nodeAPIGetter) (string, error) {
-	// Skip decoding validation for IAM tokens since they are generated with a different method
-	if settings.joinMethod != string(types.JoinMethodIAM) {
-		// This token does not need to be validated against the backend because it's not used to
-		// reveal any sensitive information. However, we still need to perform a simple input
-		// validation check by verifying that the token was auto-generated.
-		// Auto-generated tokens must be encoded and must have an expected length.
-		decodedToken, err := hex.DecodeString(settings.token)
-		if err != nil {
-			return "", trace.Wrap(err)
+	switch settings.joinMethod {
+	case string(types.JoinMethodUnspecified), string(types.JoinMethodToken), string(types.JoinMethodEC2), string(types.JoinMethodIAM):
+		if settings.joinMethod != string(types.JoinMethodIAM) {
+			decodedToken, err := hex.DecodeString(settings.token)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+			if len(decodedToken) != auth.TokenLenBytes {
+				return "", trace.BadParameter("invalid token %q", decodedToken)
+			}
 		}
 
-		if len(decodedToken) != auth.TokenLenBytes {
-			return "", trace.BadParameter("invalid token length")
+		// The provided token can be attacker controlled, so we must validate
+		// it with the backend before using it to generate the script.
+		_, err := m.GetToken(context.Background(), settings.token)
+		if err != nil {
+			return "", trace.BadParameter("invalid token")
 		}
+	default:
+		return "", trace.BadParameter("invalid join method %q", settings.joinMethod)
+	}
+
+	// We must also validate the label spec, which can be controlled by
+	// an attacker and is fed into the join script.
+	if _, err := client.ParseLabelSpec(settings.nodeLabels); err != nil {
+		return "", trace.BadParameter("invalid node-labels")
+	}
+	if strings.ContainsAny(settings.nodeLabels, "\r\n") {
```

Comment:
Part of me wants to move this logic into ParseLabelSpec but it's technically only an issue when it comes to generating the script. Thoughts?